### ISSUE TITLE
Renormalization of Shells is incorrect

### ISSUE THREAD

**Issue by wavefunction91** [evaleev/libint#46]
I've just recently bumped the libint changeset in liresearchgroup/chronusq, and the shell renormalization not longer gives the correct result. When I updated the changeset, it would no longer compile as the libint2::Shell::renorm() function has been made private (I was previously calling it explicitly). As I remembered that we (@evaleev) had spoken about this (and that @evaleev mentioned that one does not need to call this explicitly anymore), I removed the call, and compilation went smoothly. I then ran the unit tests, and every single one failed. I've narrowed it down to (at least) the renormalization of the shells:

O/STO-3G
Unnormalized Contraction Coefficients for 1S:  0.15432987 0.53532814 0.4446354
Normalized Contraction Coefficients (old) 1S:   4.251943     4.112294     1.281623
Normalized Contraction Coefficients (new) 1S: 14.112938   3.8055732    0.445048

H2O/STO-3G SCF (old) -74.9420798968
H2O/STO-3G SCF (new) -36.8001562256 (it's not half, just close)

Is there something that needs to be done viz renormalization even now?

**Comment by evaleev** [comment]
David, did the validation tests (make check) pass?

On Wednesday, January 6, 2016, David Williams-Young <
notifications@github.com> wrote:

> I've just recently bumped the libint changeset in
> liresearchgroup/chronusq, and the shell renormalization not longer gives
> the correct result. When I updated the changeset, it would no longer
> compile as the libint2::Shell::renorm() function has been made private (I
> was previously calling it explicitly). As I remembered that we (@evaleev
> <https://github.com/evaleev>) had spoken about this (and that @evaleev
> <https://github.com/evaleev> mentioned that one does not need to call
> this explicitly anymore), I removed the call, and compilation went
> smoothly. I then ran the unit tests, and every single one failed. I've
> narrowed it down to (at least) the renormalization of the shells:
>
> O/STO-3G
> Unnormalized Contraction Coefficients for 1S: 0.15432987 0.53532814
> 0.4446354
> Normalized Contraction Coefficients (old) 1S: 4.251943 4.112294 1.281623
> Normalized Contraction Coefficients (new) 1S: 14.112938 3.8055732 0.445048
>
> H2O/STO-3G SCF (old) -74.9420798968
> H2O/STO-3G SCF (new) -36.8001562256 (it's not half, just close)
>
> Is there something that needs to be done viz renormalization even now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evaleev/libint/issues/46>.
>


-- 
web: valeyev.net

**Comment by evaleev** [comment]
@wavefunction91 ping

**Comment by wavefunction91** [comment]
I do an out of source build for Libint (So i can keep track of the versions that I've moved into CQ), and that's given me come problems with the validation tests (It seems that you assume a specific location of the installed headers for the validation tests, which doesn't play nicely with the structure that I've set up). I've just repulled and am compiling everything from scratch now do see if that remidies anything, and I'm doing things in-source such that the validation tests will run (I'm not saying that an out-of-source build isn't possible, I just haven't figured out a way to get things to work properly, this is a separate issue though).

**Comment by evaleev** [comment]
by out-of-source you mean out-of-compiler-build? Such builds are tested in travis-ci, and validation tests are run there as well ... validation tests always run from _inside_ the build tree and the header locations are the same every time. Validation tests are not meant to be used outside of the build tree.

**Comment by wavefunction91** [comment]
OK, does this work with exported libraries (i.e. can one run validations with the exported library)? What I do is the following:

mkdir build && cd build
./configure <options....> --prefix=$PWD/..
make export
make install
make check

Previously (when I just pull a new changeset as opposed to repulling the entire repo), I get an error on make check because it can't find the headers for the validation. I'm not sure if this persists with a fresh clone.

**Comment by wavefunction91** [comment]
**NOTE** this validation works for a fresh clone. It could very probably be something on my end that makes it not work for pulling new changesets. I'll get back when I test this exported library with CQ

**Comment by wavefunction91** [comment]
OK, validation tests pass, but the same problem in CQ. Any ideas on what I may be able to do to isolate the problem?

**Comment by evaleev** [comment]
Supported scenarios for validation:
1. In compiler build tree: make; make check; make export
2. In exported library build tree: make; make check; make install

**Comment by evaleev** [comment]
Which compiler are you using?

**Comment by wavefunction91** [comment]
GNU 4.9.2

On Thu, Jan 7, 2016 at 1:20 PM, Eduard Valeyev <notifications@github.com>
wrote:

> Which compiler are you using?
> On Jan 7, 2016 2:25 PM, "David Williams-Young" <notifications@github.com>
> wrote:
>
> > OK, validation tests pass, but the same problem in CQ. Any ideas on what
> I
> > may be able to do to isolate the problem?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/evaleev/libint/issues/46#issuecomment-169780988>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evaleev/libint/issues/46#issuecomment-169808233>.
>



-- 
David Williams-Young
Department of Chemistry
University of Washington
Xiaosong Li Research Group

**Comment by evaleev** [comment]
I have tried with 5.2, works. Try this at the top of main, before there is any chance of overwriting memory.

```c++
auto sto3g_oxygen_1s = Shell{
        {130.709320000, 23.808861000, 6.443608300},
        {
          {0, false, {0.15432897, 0.53532814, 0.44463454}}
        },
        {{0,0,0}}
      };
std::cout << sto3g_oxygen_1s << std::endl;
```

Should see:
> Shell:( O={0,0,0}
>   {l=0,sph=0}
>  130.70932 4.25194331812225
>  23.808861 4.11229375245432
>  6.4436083 1.28162254318487

**Comment by wavefunction91** [comment]
OK, I've figured out the problem... The way I construct a my basis definition involves two steps, generation of reference shells and construction of local basis definition out of the reference, both of which have an explicit call to the standard libint2::Shell constructor (which calls libint2::Shell::renorm()). The net result being that renorm gets called twice on a given shell. Back when I originally wrote this, libint2::Shell I didn't think to look for a copy constructor, so I just built it manually. If I use the copy constructor, everything works!
